File: sdk/python-holons/holons/echo_client.py
from __future__ import annotations

"""Certification echo client for python-holons."""

import json
import os
from pathlib import Path
import sys
import time
from typing import Any, Callable, Sequence

from holons.grpcclient import dial_uri
from holons.transport import scheme

DEFAULT_SDK = "python-holons"
DEFAULT_SERVER_SDK = "go-holons"
DEFAULT_URI = "stdio://"
DEFAULT_MESSAGE = "hello"
DEFAULT_GO_BINARY = os.environ.get("GO_BIN", "go")
DEFAULT_TIMEOUT_MS = 5000


def parse_args(argv: Sequence[str]) -> dict[str, Any]:
    out: dict[str, Any] = {
        "uri": DEFAULT_URI,
        "sdk": DEFAULT_SDK,
        "server_sdk": DEFAULT_SERVER_SDK,
        "message": DEFAULT_MESSAGE,
        "go_binary": DEFAULT_GO_BINARY,
        "timeout_ms": DEFAULT_TIMEOUT_MS,
    }

    uri_set = False
    i = 0
    while i < len(argv):
        token = argv[i]
        if token == "--sdk" and i + 1 < len(argv):
            out["sdk"] = argv[i + 1]
            i += 2
            continue
        if token == "--server-sdk" and i + 1 < len(argv):
            out["server_sdk"] = argv[i + 1]
            i += 2
            continue
        if token == "--message" and i + 1 < len(argv):
            out["message"] = argv[i + 1]
            i += 2
            continue
        if token == "--go" and i + 1 < len(argv):
            out["go_binary"] = argv[i + 1]
            i += 2
            continue
        if token == "--timeout-ms" and i + 1 < len(argv):
            try:
                timeout_ms = int(argv[i + 1])
                if timeout_ms > 0:
                    out["timeout_ms"] = timeout_ms
            except ValueError:
                pass
            i += 2
            continue

        if not token.startswith("--") and not uri_set:
            out["uri"] = token
            uri_set = True

        i += 1

    return out


def build_go_echo_server_command(go_binary: str, server_sdk: str) -> list[str]:
    command = [go_binary, "run", "./cmd/echo-server", "--listen", "stdio://"]
    if server_sdk:
        command.extend(["--sdk", server_sdk])
    return command


def _go_holons_dir() -> Path:
    return Path(__file__).resolve().parents[2] / "go-holons"


def _invoke_ping(channel: Any, message: str, timeout_ms: int) -> dict[str, Any]:
    ping = channel.unary_unary(
        "/echo.v1.Echo/Ping",
        request_serializer=lambda payload: json.dumps(payload).encode("utf-8"),
        response_deserializer=lambda raw: json.loads(raw.decode("utf-8")),
    )
    timeout = max(0.1, timeout_ms / 1000.0)
    out = ping({"message": message}, timeout=timeout)
    if isinstance(out, dict):
        return out
    return {"value": out}


def run(
    argv: Sequence[str] | None = None,
    *,
    dial_uri_fn: Callable[..., Any] = dial_uri,
    now_ms: Callable[[], int] | None = None,
) -> dict[str, Any]:
    args = parse_args(argv if argv is not None else sys.argv[1:])
    now = now_ms if now_ms is not None else lambda: int(time.time() * 1000)
    started = now()

    if scheme(args["uri"]) == "stdio":
        command = build_go_echo_server_command(args["go_binary"], args["server_sdk"])
        channel = dial_uri_fn(
            args["uri"],
            stdio_command=command,
            stdio_env=dict(os.environ),
            stdio_cwd=str(_go_holons_dir()),
        )
    else:
        channel = dial_uri_fn(args["uri"])

    try:
        out = _invoke_ping(channel, str(args["message"]), int(args["timeout_ms"]))
        if out.get("message") != args["message"]:
            raise RuntimeError(f"unexpected echo message: {out.get('message')!r}")

        return {
            "status": "pass",
            "sdk": args["sdk"],
            "server_sdk": args["server_sdk"],
            "latency_ms": max(0, now() - started),
        }
    finally:
        channel.close()


def main() -> None:
    try:
        result = run()
    except Exception as exc:  # pragma: no cover - CLI guard
        print(str(exc), file=sys.stderr)
        raise SystemExit(1) from exc

    print(json.dumps(result, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: sdk/node-holons/cmd/echo-client.js
#!/usr/bin/env node
'use strict';

const path = require('node:path');
const grpc = require('@grpc/grpc-js');

const grpcclient = require('../src/grpcclient');

const DEFAULT_SDK = 'js-holons';
const DEFAULT_SERVER_SDK = 'go-holons';
const DEFAULT_URI = 'stdio://';
const DEFAULT_MESSAGE = 'hello';
const DEFAULT_GO_BINARY = process.env.GO_BIN || 'go';
const DEFAULT_TIMEOUT_MS = 5000;

const ECHO_SERVICE_DEF = {
    Ping: {
        path: '/echo.v1.Echo/Ping',
        requestStream: false,
        responseStream: false,
        requestSerialize: serialize,
        requestDeserialize: deserialize,
        responseSerialize: serialize,
        responseDeserialize: deserialize,
        originalName: 'ping',
    },
};

function parseArgs(argv) {
    const out = {
        uri: DEFAULT_URI,
        sdk: DEFAULT_SDK,
        serverSDK: DEFAULT_SERVER_SDK,
        message: DEFAULT_MESSAGE,
        goBinary: DEFAULT_GO_BINARY,
        timeoutMs: DEFAULT_TIMEOUT_MS,
    };

    for (let i = 2; i < argv.length; i += 1) {
        const token = argv[i];
        if (token === '--sdk' && i + 1 < argv.length) {
            out.sdk = argv[i + 1];
            i += 1;
            continue;
        }
        if (token === '--server-sdk' && i + 1 < argv.length) {
            out.serverSDK = argv[i + 1];
            i += 1;
            continue;
        }
        if (token === '--message' && i + 1 < argv.length) {
            out.message = argv[i + 1];
            i += 1;
            continue;
        }
        if (token === '--go' && i + 1 < argv.length) {
            out.goBinary = argv[i + 1];
            i += 1;
            continue;
        }
        if (token === '--timeout-ms' && i + 1 < argv.length) {
            const parsed = Number(argv[i + 1]);
            if (Number.isFinite(parsed) && parsed > 0) {
                out.timeoutMs = Math.trunc(parsed);
            }
            i += 1;
            continue;
        }
        if (!token.startsWith('--') && !out._uriWasSet) {
            out.uri = token;
            out._uriWasSet = true;
        }
    }

    delete out._uriWasSet;
    return out;
}

function buildGoEchoServerArgs(options = {}) {
    const args = [
        'run',
        './cmd/echo-server',
        '--listen',
        'stdio://',
    ];

    if (options.serverSDK) {
        args.push('--sdk', options.serverSDK);
    }

    return args;
}

async function run(argv = process.argv, deps = {}) {
    const args = parseArgs(argv);
    const now = typeof deps.now === 'function' ? deps.now : Date.now;
    const grpcclientModule = deps.grpcclientModule || grpcclient;

    const EchoClient = grpc.makeGenericClientConstructor(ECHO_SERVICE_DEF, 'Echo', {});
    const dialOptions = {};
    if (args.uri === 'stdio://') {
        dialOptions.command = args.goBinary;
        dialOptions.args = buildGoEchoServerArgs(args);
        dialOptions.cwd = path.resolve(__dirname, '..', '..', 'go-holons');
        dialOptions.env = process.env;
    }

    const startedAt = now();
    const session = await grpcclientModule.dialURI(args.uri, EchoClient, dialOptions);

    try {
        const out = await invokePing(session.client, args.message, args.timeoutMs);
        if (!out || out.message !== args.message) {
            throw new Error(`unexpected echo response: ${JSON.stringify(out || {})}`);
        }

        return {
            status: 'pass',
            sdk: args.sdk,
            server_sdk: args.serverSDK,
            latency_ms: Math.max(0, now() - startedAt),
        };
    } finally {
        await session.close();
    }
}

async function main() {
    try {
        const result = await run(process.argv);
        process.stdout.write(`${JSON.stringify(result)}\n`);
    } catch (err) {
        process.stderr.write(`${err.message}\n`);
        process.exit(1);
    }
}

function serialize(value) {
    return Buffer.from(JSON.stringify(value || {}));
}

function deserialize(buffer) {
    try {
        return JSON.parse(Buffer.from(buffer).toString('utf8'));
    } catch {
        return {};
    }
}

function invokePing(client, message, timeoutMs) {
    return new Promise((resolve, reject) => {
        let completed = false;
        let call = null;
        let timer = null;

        const onDone = (err, value) => {
            if (completed) return;
            completed = true;
            clearTimeout(timer);
            if (err) {
                reject(err);
                return;
            }
            resolve(value || {});
        };

        call = client.Ping({ message }, onDone);

        timer = setTimeout(() => {
            if (call && typeof call.cancel === 'function') {
                call.cancel();
            }
            onDone(new Error(`ping timeout after ${timeoutMs}ms`));
        }, timeoutMs);

        timer.unref?.();
    });
}

module.exports = {
    ECHO_SERVICE_DEF,
    buildGoEchoServerArgs,
    parseArgs,
    run,
};

if (require.main === module) {
    main();
}
